File: clock-gen-async.pl
#!/usr/bin/env perl

# Clock a MIDI device, like a drum machine or sequencer.
# Example: perl clock-gen-async.pl usb 90

use v5.36;
use feature 'try';
use IO::Async::Loop ();
use IO::Async::Timer::Periodic ();
use MIDI::RtMidi::FFI::Device ();

my $name = shift || 'usb'; # MIDI sequencer device
my $bpm  = shift || 120; # beats per minute

my $interval = 60 / $bpm / 24; # seconds / bpm / clocks-per-beat

# open the named midi device for output
my $midi_out = RtMidiOut->new;
try { # this will die on windows but is needed for mac
    $midi_out->open_virtual_port('RtMidiOut');
}
catch ($e) {}
$midi_out->open_port_by_name(qr/\Q$name/i);

$midi_out->start; # start the sequencer

$SIG{INT} = sub { # halt gracefully
    say "\nStop";
    try {
        $midi_out->panic; # make sure all notes are off
        $midi_out->stop; # stop the sequencer
    }
    catch ($e) {
        warn "Can't halt the MIDI out device: $e\n";
    }
    exit;
};

my $loop = IO::Async::Loop->new;

my $timer = IO::Async::Timer::Periodic->new(
   interval => $interval,
   on_tick  => sub { $midi_out->clock }, # send a clock tick!
);
$timer->start;

$loop->add($timer);
$loop->run;

File: clock-gen-async.py
import asyncio
import mido
from mido import Message
import sys

def periodic_task(loop, interval, task_func):
    task_func()
    # Schedule the next call recursively
    loop.call_later(interval, periodic_task, loop, interval, task_func)

def clock():
    global midi_out
    midi_out.send(Message('clock'))
    print("MIDI clock sent")

async def main(interval):
    loop = asyncio.get_running_loop()
    # delay, callback, *args passed to the callback
    loop.call_later(interval, periodic_task, loop, interval, clock)
    await asyncio.sleep(10)  # Keep main alive to see it work

if __name__ == '__main__':
    name = sys.argv[1] if len(sys.argv) > 1 else 'USB MIDI Interface'
    bpm  = int(sys.argv[2]) if len(sys.argv) > 2 else 120

    clock_interval = 60 / bpm / 24

    with mido.open_output(name) as midi_out:
        midi_out.send(Message('start'))
        try:
            asyncio.run(main(clock_interval))
        except KeyboardInterrupt:
            midi_out.send(Message('stop'))
            print("\nStop")
